### PR TITLE
Application Pipelines

### LINKED ISSUE #2075: Application > Pipelines
### Epic

#1492

### Description

**As a:** FlowForge Team Owner

**I want to:** configure multiple instances as a single "pipeline" in FlowForge

**So that:** I can experiment in development without fear of compromising production environments, and then easily roll our when I choose to.

### Which customers would this be availble to

Licensed Edition (EE)

### Acceptance Criteria

- [x] #2082
- [x] #2076
- [x] #2248

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### LINKED ISSUE #2076: Enable configuration of "pipelines" at the API and DB level
### Story

https://github.com/flowforge/flowforge/issues/2075

### Description

We need to add a new "Pipeline" model to the DB. This should enable definition of multiple stages within a single pipeline. Discussion should be undertaken to define what features need to be stored at the "stage" level too.

#### 1. Pipelines

We have a single "Pipeline" object which stores everything in a single object

#### 2. Pipelines & Stages

We have a "child" model called a "Stage" which maps the relevant data on a stage-by-stage basis. Then, the 'Pipeline" object is simply a mapping of multiple "Stages"

### LINKED ISSUE #2082: New /application/pipelines view
### Story

https://github.com/flowforge/flowforge/issues/2075

### Description

Screenshot 2023-05-03 at 14 12 40

#### Acceptance Criteria
- [x] User can create a new pipeline
- [x] User can add stages to the pipeline
- [x] User can manually kick off a given stage of the pipeline
- [ ] User can remove a stage from a pipeline

### PR DESCRIPTION
**Pull request by joepavitt**
## Description

Introduces a new "Pipelines" view at the Application level.

Functionality:
- Can add a pipeline
- Can delete a pipeline
- Can add a stage to a pipeline and define which instance that stage links to
- Can manually run a single stage of a pipeline (iteration 1 - future would be fully automated pipeline)

#### To-do:

- [ ] Test Coverage:
    - [ ]  Server-side
    - [x]  E2E Frontend
- [x] Polling of live status & "last deployed" for instances in a stage
- [x] Audit Logging (Added in separate PR for easier reviewing: https://github.com/flowforge/flowforge/pull/2098)
- [x] PostHog Events
- [ ] Improved accuracy of "last updated" - seems to change when a project is added to a pipeline? Would ideally be last restart/deployment of the nodes
- [ ] Check `needsPermission` for new endpoints

## Related Issue(s)

Closes #2082
Closes #2076 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts

### CODE REVIEW

**Comment by joepavitt** [comment]
> Polling of live status & "last deployed" for instances in a stage

@Pezmc would appreciate some advice/input on this outstanding task. We still just repeating a HTTP request every X seconds, or did something more sophisticated get added that I haven't seen?

**Comment by joepavitt** [comment]
Development work _should_ now be complete, apart from:

- Improved accuracy of "last updated" - seems to change when a project is added to a pipeline? Would ideally be last restart/deployment of the nodes
- Check needsPermission for new endpoints (@knolleary need your input here - can connect early next week)

For early next week, need to prioritise test coverage for the new work. I'm happy to take the E2E front-end testing, but will need assistance server-side. @Pezmc if you have the bandwidth, would be great to get the assist.

TLDR for server-side changes and where test coverage is likely required:
- New API endpoints (`/applications/:applicationid/pipelines` and `/pipelines/:pipelineid/:stages`)
- New models, views & controllers: `Pipeline` and `PipelineStage`

**Comment by joepavitt** [comment]
Have added limited E2E testing as currently, we have no way of testing licensed features in the E2E tests - also the same issue with the Team Library.

**Comment by knolleary** [comment]
The unit test failure on postgres comes down to how CE/EE features co-exist in the database models and when/if migrations get applied.

For a clean run of the forge app (no existing database), the code relies 100% on the sequelize models to generate the database structure. The migrations do not get used.

This PR introduces a migration that adds a relation between the (CE) Projects table and (EE) PipelineStages table. However, in a clean run, that migration will not get applied. So the Projects table does not have the expected PipelineStageId column.

Furthermore, the existing migration will *fail* when upgrading an existing instance of FF - because the migrations run before sequelize initiates the models, so the PipelineStages table won't exist at that point in time.

I need to spend some more time working out how to fix all of this.

In summary it involves:

1. Adding migrations to create the new tables
2. Consider adding a through table relation between Project and PipelineStage (like we have Team -> TeamMember -> User). This will remove the need to modify the Projects table at all.

This is a few hours work, including tests.

**Comment by knolleary** [comment]
Having been poking at this to understand the database requirements a bit better. Hit a couple UX things:

1. Bug: When creating a pipeline stage it is possible to pick an instance that is already assigned to a stage. If you do that, the pipeline view no longer renders the pipeline - so you cannot edit/delete it. Had to delete the pipeline/stages in the database to get things back.
    Recreate:
        1. Create pipeline
        2. Create first stage with Instance A
        3. Create second stage, pick Instance A again
        4. The pipeline view is now blank - no way to add/edit/delete the pipeline
2. Missing feature: No ability to edit anything - if you make a typo in a name or pick the wrong instance you have to delete the whole pipeline and start again

**Comment by Pezmc** [comment]
I started work on fixing 1, but have been held up with failing migrations locally. Am starting work on 1. Adding migrations to create the new tables from the [previous comment](https://github.com/flowforge/flowforge/pull/2094#issuecomment-1539960498).

File: forge/ee/routes/pipeline/index.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    }, async (request, reply) => {
+        const team = await request.teamMembership.getTeam()
+        const name = request.body.name?.trim() // name of the stage
+        const instanceId = request.body.instance // instance id
+
+        let stage
+        try {
+            const options = {
+                name,
+                instance: instanceId
```
Comment (Pezmc) [inline forge/ee/routes/pipeline/index.js:58]: Please always include ID in option name if it's expecting an ID rather than a whole object.

```suggestion
                instanceId: instanceId
```

File: frontend/src/pages/application/PipelineStage/edit.vue [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    async mounted () {
+        this.mounted = true
+        this.loadInstances()
+    },
+    methods: {
+        async create () {
+            this.loading = true
+            const options = {
+                name: this.input.name,
+                instance: this.input.instance
```
Comment (Pezmc) [inline frontend/src/pages/application/PipelineStage/edit.vue:129]:
```suggestion
                instanceId: this.input.instance
```

File: frontend/src/api/pipeline.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+/**
+ * @param {string} pipelineId
+ * @param {object} stage
+ */
+const addPipelineStage = async (pipelineId, stage) => {
+    const options = {
+        name: stage.name,
+        instance: stage.instance
```
Comment (Pezmc) [inline frontend/src/api/pipeline.js:24]:
```suggestion
        instanceId: stage.instance
```

**Comment by joepavitt** [comment]
> Missing feature: No ability to edit anything - if you make a typo in a name or pick the wrong instance you have to delete the whole pipeline and start again

Not entirely an accident. Given the size of the task at hand already, and how close it was getting to release, I made a call to not add that in for now, as there would be a lot of complexities around deleting stages in particular (e.g. then having to remap targets from adjacent stages, etc.). In the interest of iteration, was happy to ship as is (it takes less than a minute to delete the pipeline and re-create) so I was okay with that.

**Comment by Pezmc** [comment]
Tracking of work needed to consider this PR "ready" moved to https://github.com/flowforge/flowforge/issues/2124. Will be branching from `2075-application-pipelines` and using this PR as a feature branch target for any PRs.

**Review by Pezmc — APPROVED** [review]
Approving for myself, see: https://flowforgeworkspace.slack.com/archives/C032Q63FGG1/p1685094359788189

**Review by Pezmc — APPROVED** [review]
Approving for myself, see: https://flowforgeworkspace.slack.com/archives/C032Q63FGG1/p1685094359788189
